### PR TITLE
Dynamic library loader behaviour/plan/runtimecondition

### PR DESCRIPTION
**Pull request by triccyx**
Dynamic library loader behaviour/plan/runtimecondition

Wrike task: https://www.wrike.com/open.htm?id=936029023
Doc: https://docs.google.com/document/d/1edam2pjbACQxJqRFaZazQUvYRG56vGJ8K6mXmkFpQEk/edit#

### CODE REVIEW

File: alica_dynamic_loading/include/DynamicLoadingUtils.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    while (std::getline(ssToTokenize, tmp, delimter)) {
+        tokens.push_back(tmp);
+    }
+
+    return tokens;
+}
+
+inline std::string calculateLibraryPath()
+{
+    const char* ldLibraryPath = std::getenv("LD_LIBRARY_PATH");
```
Comment (veerajsk) [inline alica_dynamic_loading/include/DynamicLoadingUtils.h]: can return null pointer, so need to check that right?

File: alica_dynamic_loading/include/DynamicLoadingUtils.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    auto tokens = tokenizeStr(ldLibraryPath, ':');
+    if (tokens.size() < 1) {
+        std::cerr << "Error:"
```
Comment (veerajsk) [inline alica_dynamic_loading/include/DynamicLoadingUtils.h]: should use alica's logger to log errors

File: alica_dynamic_loading/src/DynamicBehaviourCreator.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+namespace alica
+{
+
+DynamicBehaviourCreator::DynamicBehaviourCreator() {}
+
+DynamicBehaviourCreator::~DynamicBehaviourCreator() {}
+
+std::unique_ptr<BasicBehaviour> DynamicBehaviourCreator::createBehaviour(int64_t behaviourId, BehaviourContext& context)
+{
+    std::string libraryPath = calculateLibraryPath();
```
Comment (veerajsk) [inline alica_dynamic_loading/src/DynamicBehaviourCreator.cpp]: this can be done just once during initialization

File: alica_dynamic_loading/include/DynamicLoadingUtils.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                  << "Missing LD_LIBRARY_PATH" << std::endl;
+        return "";
+    }
+
+    std::cerr << "Debug:"
+              << "use library path from LD_LIBRARY_PATH:" << tokens[0] << std::endl;
+
+    return tokens[0];
```
Comment (veerajsk) [inline alica_dynamic_loading/include/DynamicLoadingUtils.h]: need to consider all tokens to find the library. Usually LD_LIBRARY_PATH has multiple paths defined in it

File: alica_dynamic_loading/src/DynamicConditionCreator.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            context.name,                                               // symbol to import
+            boost::dll::load_mode::append_decorations                   // do append extensions and prefixes
+    );
+
+    std::shared_ptr<BasicCondition> createdCondition = _conditionCreator(context);
+
+    return createdCondition;
+}
+
+std::shared_ptr<BasicCondition> DynamicConditionCreator::createConditions(int64_t conditionConfId)
```
Comment (veerajsk) [inline alica_dynamic_loading/src/DynamicConditionCreator.cpp]: what's this overload for?

File: alica_engine/src/engine/expressionhandler/ExpressionHandler.cpp [hunk 6]
```diff
@@ -37,15 +38,17 @@ void ExpressionHandler::attachAll(PlanRepository& pr, AlicaCreators& creatorCtx)
 
         if (p->getPreCondition() != nullptr) {
             if (p->getPreCondition()->isEnabled()) {
-                p->_preCondition->setBasicCondition(creatorCtx.conditionCreator->createConditions(p->getPreCondition()->getId()));
+                ConditionContext context{p->getPreCondition()->getName(), p->getPreCondition()->_libraryName, p->getPreCondition()->getId()};
```
Comment (veerajsk) [inline alica_engine/src/engine/expressionhandler/ExpressionHandler.cpp]: isn't `_library_name` private? can't you use `getLibraryName()` instead?

File: alica_engine/src/engine/model/Condition.cpp [hunk 7]
```diff
@@ -76,4 +76,9 @@ void Condition::setBasicConstraint(const std::shared_ptr<BasicConstraint>& basic
     _basicConstraint = basicConstraint;
 }
 
+std::string Condition::getLibraryName()
```
Comment (veerajsk) [inline alica_engine/src/engine/model/Condition.cpp]: can make it a const method

**Review by veerajsk — COMMENTED** [review]
only reviewed the engine changes

File: alica_dynamic_loading/include/DynamicLoadingUtils.h (hunk 1 above)
Comment (triccyx) [reply alica_dynamic_loading/include/DynamicLoadingUtils.h]: Yes of course

File: alica_dynamic_loading/src/DynamicConditionCreator.cpp (hunk 5 above)
Comment (triccyx) [reply alica_dynamic_loading/src/DynamicConditionCreator.cpp]: leftover

File: alica_engine/src/engine/model/Condition.cpp (hunk 7 above)
Comment (triccyx) [reply alica_engine/src/engine/model/Condition.cpp]: ok

File: alica_engine/src/engine/expressionhandler/ExpressionHandler.cpp (hunk 6 above)
Comment (triccyx) [reply alica_engine/src/engine/expressionhandler/ExpressionHandler.cpp]: ok

File: alica_dynamic_loading/src/DynamicBehaviourCreator.cpp (hunk 3 above)
Comment (triccyx) [reply alica_dynamic_loading/src/DynamicBehaviourCreator.cpp]: ok

File: alica_dynamic_loading/include/DynamicLoadingUtils.h (hunk 2 above)
Comment (triccyx) [reply alica_dynamic_loading/include/DynamicLoadingUtils.h]: ok

File: alica_dynamic_loading/include/DynamicLoadingUtils.h (hunk 4 above)
Comment (triccyx) [reply alica_dynamic_loading/include/DynamicLoadingUtils.h]: I use the first folder because of the setup.sh script in install creates ad add at the start of LD_LIBRARY_PATH.
However to be bug-free as much as possible I will search in each folder of LD_LIBRARY_PATH for the library

Comment (Dale-Koenig) [reply alica_dynamic_loading/include/DynamicLoadingUtils.h]: You cannot guarantee that the LD_LIBRARY_PATH will not be further modified, even if you initially prepend

Comment (triccyx) [reply alica_dynamic_loading/include/DynamicLoadingUtils.h]: Yes, my only doubt about this approach is that we risk loading a library with the same name but in the wrong folder. Maybe an old version. 
Perhaps we should also add a version tag in the .beh/.pml/.cnd files and match with a number in the compiled library.
```xml
...
  "libraryName": "alica_turtlesim_library",
  "libraryVersion": 1001
}
```

File: supplementary/alica_ros_turtlesim/include/alica_ros_turtlesim/base.hpp [hunk 8]
```diff
@@ -10,13 +10,18 @@ class Base
...
     void start();
 
 private:
     ros::AsyncSpinner spinner;
     alica::AlicaContext* ac;
+    void ALICATurtleWorldModelCallInit(ros::NodeHandle& nh, ros::NodeHandle& priv_nh);
+
+    void ALICATurtleWorldModelCallDel();
+
+    bool _doDynamic{false};
```
Comment (triccyx) [reply supplementary/alica_ros_turtlesim/include/alica_ros_turtlesim/base.hpp]: OK

File: supplementary/alica_ros_turtlesim/launch/turtle.launch [hunk 9]
```diff
@@ -6,6 +6,7 @@
     <arg name="master_plan"    default="Master" />
     <arg name="roleset"        default="Roleset" />
     <arg name="alica_path"     default="$(find alica_ros_turtlesim)/alica" />
+    <arg name="dynamic"        default="NO" />
```
Comment (triccyx) [reply supplementary/alica_ros_turtlesim/launch/turtle.launch]: OK

File: supplementary/alica_ros_turtlesim/src/base.cpp [hunk 10]
```diff
@@ -13,42 +18,79 @@
...
 }
 
 Base::~Base()
 {
     spinner.stop(); // stop spinner before terminating engine
     ac->terminate();
     delete ac;
-    ALICATurtleWorldModel::del();
+    if (_doDynamic)
+        ALICATurtleWorldModel::del();
```
Comment (triccyx) [reply supplementary/alica_ros_turtlesim/src/base.cpp:117]: I can't this call is just the Singleton ALICATurtleWorldModel deallocation.
However, here there is a bug should be:
```
if (!_doDynamic)
```
also is missing the `_doDynamic` part

File: alica_dynamic_loading/include/DynamicLoadingUtils.h [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+#include <filesystem>
+#include <iostream>
+#include <sstream>
+#include <stack>
+#include <string>
+#include <vector>
+
+namespace alica
+{
+
```
Comment (Dale-Koenig) [inline alica_dynamic_loading/include/DynamicLoadingUtils.h:14]: I would put these in cpp file instead of header

File: supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp [hunk 12]
```diff
@@ -15,7 +15,7 @@ namespace alica
 bool conditionMove2Init748720375848597116(const Blackboard* input, const RunningPlan* rp, const IAlicaWorldModel* wm)
 {
     /*PROTECTED REGION ID(condition748720375848597116) ENABLED START*/
-    return turtlesim::ALICATurtleWorldModel::get()->getInit();
+    return true; // turtlesim::ALICATurtleWorldModel::get()->getInit(); todo luca when ITransitionConditionCreator can be written dynamic
```
Comment (Dale-Koenig) [inline supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: What is going on here?

File: supplementary/alica_ros_turtlesim/launch/turtle.launch (hunk 9 above)
Comment (Dale-Koenig) [inline supplementary/alica_ros_turtlesim/launch/turtle.launch]: Make this a bool (true or false) and simplify code logic accordingly

File: supplementary/alica_ros_turtlesim/include/alica_ros_turtlesim/base.hpp (hunk 8 above)
Comment (Dale-Koenig) [inline supplementary/alica_ros_turtlesim/include/alica_ros_turtlesim/base.hpp]: _loadDynamically

File: supplementary/alica_ros_turtlesim/src/base.cpp (hunk 10 above)
Comment (Dale-Koenig) [inline supplementary/alica_ros_turtlesim/src/base.cpp:117]: Consider putting cleanup in the corresponding destructor

Comment (triccyx) [reply supplementary/alica_ros_turtlesim/src/base.cpp:117]: Fixed

File: supplementary/supplementary_tests/alica_supplementary_test_library/src/WaitBehaviour.cpp [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+    // if (!getBlackboard()) {
+    //     std::cerr << "Blackboard null" << std::endl;
+    // }
+    //  LockedBlackboardRW bb(*(getBlackboard()));
+    /*
+    int myTime=bb.get<int>("masterKey");
+    std::cerr<<"Wait for:"<<myTime<<std::endl;
+    */
+    std::this_thread::sleep_for(std::chrono::seconds(myTime));
```
Comment (Dale-Koenig) [inline supplementary/supplementary_tests/alica_supplementary_test_library/src/WaitBehaviour.cpp]: I guess if this is in the test package we don't need to make it useful, so just do whatever but don't leave commented code

File: supplementary/supplementary_tests/src/test/test_alica_dynamicloadcreator.cpp [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Create behaviour form dll
+    IAlicaWorldModel wm;
+    auto creator = std::make_unique<alica::DynamicBehaviourCreator>();
+
+    BehaviourContext ctx{&wm, behaviourModel->getName(), behaviourModel, nullptr};
+    std::unique_ptr<BasicBehaviour> behaviour = creator->createBehaviour(10, ctx);
+
+    ASSERT_EQ("waitbehaviour", behaviour->getName());
+    // behaviour->start(nullptr);
```
Comment (Dale-Koenig) [inline supplementary/supplementary_tests/src/test/test_alica_dynamicloadcreator.cpp]: ?

File: supplementary/supplementary_tests/alica_supplementary_test_library/src/WaitBehaviour.cpp (hunk 13 above)
Comment (triccyx) [reply supplementary/supplementary_tests/alica_supplementary_test_library/src/WaitBehaviour.cpp]: ok

File: supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp (hunk 12 above)
Comment (triccyx) [reply supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: The only clear way to get rid of this is to implement also for TransitionConditionCreator a dynamic version.
And create a copy of these files related to TransitionCondition in the turtlesim dynamic library.
I proceed in this direction but it took me a while.

File: supplementary/supplementary_tests/src/test/test_alica_dynamicloadcreator.cpp (hunk 14 above)
Comment (triccyx) [reply supplementary/supplementary_tests/src/test/test_alica_dynamicloadcreator.cpp]: removed

**Review by Dale-Koenig — APPROVED** [review]

File: supplementary/alica_ros_turtlesim/README.md [hunk 15]
```diff
@@ -761,6 +761,18 @@ Run application with roslaunch. video
 - Start moving.
   `rostopic pub /init std_msgs/Empty "{}" `
 
+### 8.3 Run with Dynamic loading
+
+Is it possible to run Turtlesim using a dynamic library that contains all the Plan/Behaviour/Runtimecondition with:
```
Comment (veerajsk) [inline supplementary/alica_ros_turtlesim/README.md]:
```suggestion
It is possible to run Turtlesim using a dynamic library that contains all the Plan/Behaviour/Runtimecondition with:  
```

File: supplementary/alica_ros_turtlesim/alica_turtlesim_library/include/turtle.hpp [hunk 16]
```diff
@@ -0,0 +1,40 @@
+#ifndef ALICA_TURTLE_SIM_TURTLE_HPP
+#define ALICA_TURTLE_SIM_TURTLE_HPP
+
+#include <ros/ros.h>
+#include <turtlesim/Pose.h>
+
+namespace turtlesim
+{
+/*
+    ALICATurtleWoroldModel
```
Comment (veerajsk) [inline supplementary/alica_ros_turtlesim/alica_turtlesim_library/include/turtle.hpp]:
```suggestion
    ALICATurtleWorldModel
```

File: supplementary/alica_ros_turtlesim/src/base.cpp [hunk 17]
```diff
@@ -13,42 +18,103 @@
...
     ac->setLogger<alicaRosLogger::AlicaRosLogger>(agent_id);
+
+    // create world model
+    if (_loadDynamically) {
+        ALICATurtleWorldModelCallInit(nh, priv_nh);
+    } else
+        ALICATurtleWorldModel::init(nh, priv_nh);
+}
+
+void Base::ALICATurtleWorldModelCallInit(ros::NodeHandle& nh, ros::NodeHandle& priv_nh)
```
Comment (veerajsk) [inline supplementary/alica_ros_turtlesim/src/base.cpp:48]: We need to get rid of this singleton for the world model & use `AlicaContext::setWorldModel<>()` instead & the behaviours should access the world model from the engine. I think we need to rework this API to either load the world model dynamically or statically create it as it is done now. 
Can go in a separate PR (need to create a wrike task for it)

File: supplementary/alica_ros_turtlesim/src/base.cpp [hunk 18]
```diff
@@ -13,42 +18,103 @@
...
 {
     spinner.stop(); // stop spinner before terminating engine
     ac->terminate();
     delete ac;
-    ALICATurtleWorldModel::del();
+
+    if (_loadDynamically) {
+        ALICATurtleWorldModelCallDel();
+    } else {
+        ALICATurtleWorldModel::del();
```
Comment (veerajsk) [inline supplementary/alica_ros_turtlesim/src/base.cpp:117]: many places with missing braces for single line statements

**Review by veerajsk — APPROVED** [review]

File: supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp [hunk 19]
```diff
@@ -15,7 +15,7 @@ namespace alica
 bool conditionMove2Init748720375848597116(const Blackboard* input, const RunningPlan* rp, const IAlicaWorldModel* wm)
 {
     /*PROTECTED REGION ID(condition748720375848597116) ENABLED START*/
-    return turtlesim::ALICATurtleWorldModel::get()->getInit();
+    return true; // turtlesim::ALICATurtleWorldModel::get()->getInit(); // todo luca when ITransitionConditionCreator can be written dynamic
```
Comment (veerajsk) [inline supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: so the conditions are always statically linked now (even if dynamic=true)? Are you planning to load them dynamically in this PR?

Comment (triccyx) [reply supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: Yes, I have written and tested the DynamicTransitionCreator. I need to change turtlesim demo.

Comment (triccyx) [reply supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: done

File: supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp (hunk 12 above)
Comment (triccyx) [reply supplementary/alica_ros_turtlesim/alica/Expr/src/conditions/conditions.cpp]: done
